src/gowrapper/ring/ring.go

```diff
@@ -85,3 +109,197 @@ func lattigo_readNewFromSampler(samplerHandle Handle14) Handle14 {
+func lattigo_copyPolyAtLevel(dstPolyHandle Handle14, dstIndex uint64, srcPolyHandle Handle14, srcIndex uint64) {
+	src := getStoredPoly(srcPolyHandle)
+	dst := getStoredPoly(dstPolyHandle)
+	srcIdx := int(srcIndex)
+	dstIdx := int(dstIndex)
+	copy(dst.Coeffs[dstIdx], src.Coeffs[srcIdx])
+
+	// copy(dst.Buff[dst.N()*dstIdx:dst.N()*(dstIdx+1)], src.Buff[src.N()*srcIdx:src.N()*(srcIdx+1)])
+
+}
+
+//export lattigo_polyDegree
+func lattigo_polyDegree(polyHandle Handle14) uint64 {
+	poly := getStoredPoly(polyHandle)
+	return uint64(poly.N())
+}
+
+//export lattigo_N
+func lattigo_N(ringHandle Handle14) uint64 {
+	ring := getStoredRing(ringHandle)
+	return uint64(ring.N)
+}
+
+//export lattigo_permuteNTTIndex
+func lattigo_permuteNTTIndex(ringHandle Handle14, galEl uint64, outValues *C.constULong) {
+	ring := getStoredRing(ringHandle)
+	res := ring.PermuteNTTIndex(galEl)
+	size := unsafe.Sizeof(uint64(0))
+	basePtr := uintptr(unsafe.Pointer(outValues))
+	for i := range res {
+		*(*uint64)(unsafe.Pointer(basePtr + size*uintptr(i))) = res[i]
+	}
+}
+
+//export lattigo_permuteNTTWithIndexLvl
+func lattigo_permuteNTTWithIndexLvl(ringHandle Handle14, level uint64, polyInHandle Handle14, index *C.constULong, polyOutHandle Handle14) {
+	ring := getStoredRing(ringHandle)
+	polyIn := getStoredPoly(polyInHandle)
+	polyOut := getStoredPoly(polyOutHandle)
+	Index := make([]uint64, ring.N)
+	indexPtr := uintptr(unsafe.Pointer(index))
+	size := unsafe.Sizeof(uint64(0))
+	for i := range Index {
+		Index[i] = *(*uint64)(unsafe.Pointer(indexPtr + size*uintptr(i)))
+	}
+	ring.PermuteNTTWithIndexLvl(int(level), polyIn, Index, polyOut)
+}
+
+//export lattigo_log2OfInnerSum
+func lattigo_log2OfInnerSum(levelQ int, ringQHandle, polyHandle Handle14) int {
+	ringQ := getStoredRing(ringQHandle)
+	poly := getStoredPoly(polyHandle)
+	return ringQ.Log2OfInnerSum(levelQ, poly)
+}
+
+//export lattigo_MulCoeffsMontgomeryAndAddLvl
+func lattigo_MulCoeffsMontgomeryAndAddLvl(ringQPHandle Handle14, levelQ, levelP int, p1Handle, p2Handle, p3Handle Handle14) {
+	p1 := getStoredPolyQP(p1Handle)
+	p2 := getStoredPolyQP(p2Handle)
+	p3 := getStoredPolyQP(p3Handle)
+	ringQP := getStoredRingQP(ringQPHandle)
+	ringQP.RingQ.MulCoeffsMontgomeryAndAddLvl(levelQ, p1.Q, p2.Q, p3.Q)
+	ringQP.RingP.MulCoeffsMontgomeryAndAddLvl(levelP, p1.P, p2.P, p3.P)
+}
+
+//export lattigo_MulCoeffsMontgomeryAndAddLvlRing
+func lattigo_MulCoeffsMontgomeryAndAddLvlRing(ringHandle Handle14, level int, p1Handle, p2Handle, p3Handle Handle14) {
+	p1 := getStoredPoly(p1Handle)
+	p2 := getStoredPoly(p2Handle)
+	p3 := getStoredPoly(p3Handle)
+	ring := getStoredRing(ringHandle)
+	ring.MulCoeffsMontgomeryAndAddLvl(level, p1, p2, p3)
+}
+
+//export lattigo_equals
+func lattigo_equals(p1Handle, p2Handle Handle14) int {
+	p1 := getStoredPoly(p1Handle)
+	p2 := getStoredPoly(p2Handle)
+	if p1.Equals(p2) {
+		return 1
+	} else {
+		return 0
+	}
+}
+
+//export lattigo_printPoly
+func lattigo_printPoly(pHandle Handle14) {
```

Comment:
I think it will be cleaner to add a method for retrieving the coefficients (and then we can print them in the cpp side)